game-core/src/main/java/games/strategy/triplea/ui/PoliticsPanel.java

```diff
@@ -57,61 +55,81 @@
         SwingAction.of(
             "Do Politics...",
             e -> {
-              final Dimension screenResolution = Toolkit.getDefaultToolkit().getScreenSize();
-              final int availHeight = screenResolution.height - 96;
-              final int availWidth = screenResolution.width - 30;
-              final int availHeightOverview = (int) ((float) availHeight * 2 / 3);
-              final int availHeightChoice = (int) ((float) availHeight / 3);
-
               final JDialog politicalChoiceDialog = new JDialog(parent, "Political Actions", true);
               final Insets insets = new Insets(1, 1, 1, 1);
               final JPanel politicalChoicePanel = new JPanel();
               politicalChoicePanel.setLayout(new GridBagLayout());
               final PoliticalStateOverview overview =
-                  new PoliticalStateOverview(getData(), getMap().getUiContext(), false);
-              final JScrollPane overviewScroll = new JScrollPane(overview);
-              overviewScroll.setBorder(BorderFactory.createEmptyBorder());
-              // add 26 to height when the actions are empty, because for some stupid reason java
-              // calculates the pack size wrong
-              // (again)...
-              // add 20 to either when the opposite needs scroll bars, because that is how big
-              // scroll bars are..
-              overviewScroll.setPreferredSize(
-                  new Dimension(
-                      (overviewScroll.getPreferredSize().width > availWidth
-                          ? availWidth
-                          : (overviewScroll.getPreferredSize().width
-                              + (overviewScroll.getPreferredSize().height > availHeightOverview
-                                  ? 20
-                                  : 0))),
-                      (overviewScroll.getPreferredSize().height > availHeightOverview
-                          ? availHeightOverview
-                          : (overviewScroll.getPreferredSize().height
-                              + (validPoliticalActions.isEmpty() ? 26 : 0)
-                              + (overviewScroll.getPreferredSize().width > availWidth ? 20 : 0)))));
+                  new PoliticalStateOverview(getData(), getMap().getUiContext(), false) {
+
+                    private GridBagConstraints makeGridBagConstraints(
+                        final Insets insets, final int row) {
+                      return new GridBagConstraints(
+                          1,
+                          row,
+                          1,
+                          1,
+                          1.0,
+                          1.0,
+                          GridBagConstraints.WEST,
+                          GridBagConstraints.HORIZONTAL,
+                          insets,
+                          0,
+                          0);
+                    }
+
+                    @Override
+                    protected JPanel getRelationshipComponent(
```

Comment:
Overriding concrete methods, general usage of inheritance can create a number of problems. For example, in the `PoliticalStateOverview` class, it's not going to be clear we have dependencies on any method that is called by this override method. It's a 'spooky action at distance', where changes in `PoliciticalStateOverview` impact this class.

A good rule of thumb is to simply avoid overriding concrete types. I think this could potentially be fixed by extracting the common components into an abstract class. `getRelationshipComponent` would then be an abstract method, and there would be two child classes, `PoliticalStateOverview` and `PoliticsPanel`. This way the common code is in an abstract class and because it is abstract, it's more clear that is shared code. Hopefully that seems like a good design idea, creates a more tree like structure where children add to functionality of the parent rather than changing the behavior of the parent. 

TL:DR, I'd like to see us avoid an override of a concrete method. It can create a lot of complexity/confusion and breaks modularity. The common code could also be extracted to another class and we could use composition, so that the common code is then invoked instead of being inherited. eg:
```
doRendering() {
  RelationshipType relType = relationshipCalculator.calculateRelationship();
  return getRelationshipComponent(relType);
}
```
Given the existing structure, the abstract base class might be the easier route.  @asvitkine any thoughts on this one?